lib/dal/src/component.rs

```diff
@@ -1399,10 +1462,435 @@ impl Component {
+                if let Some(parent) = Self::get_parent_by_id(ctx, working_component_id).await? {
+                    work_queue.push_back(parent);
+                }
+            }
+        }
+
+        Ok(found_sockets)
+    }
+    /// Finds all inferred connections for the [`Component`]
+    /// A connection is inferred if it's input or output sockets are being driven
+    /// as a result of parentage (for example, dropping a [`Component`] in another [`Component`]
+    /// that is either a [`ComponentType::ComponentFrameUp`] or [`ComponentType::ComponentFrameDown`])
+    #[instrument(level = "debug", skip(ctx))]
+    pub async fn inferred_connections(
+        &self,
+        ctx: &DalContext,
+    ) -> ComponentResult<Vec<InferredIncomingConnection>> {
+        let to_component_id = self.id();
+        let mut connections = vec![];
+        let input_sockets =
+            Self::input_socket_attribute_values_for_component_id(ctx, to_component_id).await?;
+        for (to_input_socket_id, input_socket_match) in input_sockets.into_iter() {
+            if let Some(output_socket_match) =
+                Self::find_inferred_connection_to_input_socket(ctx, input_socket_match).await?
+            {
+                let implicit_edge = InferredIncomingConnection {
+                    to_component_id,
+                    to_input_socket_id,
+                    from_component_id: output_socket_match.component_id,
+                    from_output_socket_id: output_socket_match.output_socket_id,
+                };
+                debug!("Found inferred edge: {:?}", implicit_edge);
+                connections.push(implicit_edge);
+            }
+        }
+        Ok(connections)
+    }
+
+    /// For the provided [`InputSocketMatch`], find the first [`OutputSocketMatch`] that should
+    /// drive this [`InputSocket`] by searching down the descendants of the [`Component`],
+    /// checking children first and walking down until we find a single match
+    /// Note: If we find multiple matches (for example, multiple children of a
+    /// [`ComponentType::ComponentFrameUp`], we return [`None`],
+    /// forcing the user to make an explicit, manual connection, by drawing an edge)
+    #[instrument(level = "debug", skip(ctx))]
+    async fn find_first_output_socket_match_in_descendants(
+        ctx: &DalContext,
+        input_socket_match: InputSocketMatch,
+        component_types: Vec<ComponentType>,
+    ) -> ComponentResult<Option<OutputSocketMatch>> {
+        let mut found_match: Option<OutputSocketMatch> = None;
+        let component_id = input_socket_match.component_id;
+        let children = Component::get_children_for_id(ctx, component_id).await?;
+        //load up the children and look for matches
+        let mut work_queue: VecDeque<Vec<ComponentId>> = VecDeque::new();
+        work_queue.push_front(children);
+
+        'parents: while let Some(children) = work_queue.pop_front() {
+            for child_component in children {
```

Comment:
Generally speaking, using a named loop is a bit of a smell that the logic is too complex for one function (cc @zacharyhamm since I think you've mentioned this before). I'd consider breaking this out into another function for processing each set of children. The return type could influence how you break out of the `while` loop.